File replace becomes hardlink friendly

lorengordon opened the pull request:
Fixes #20969.

This PR replaces https://github.com/saltstack/salt/pull/20989.

This commit introduces the parameter preserve_inode to control the
behavior of file.replace. If True, the file is copied, then
file.replace writes to the original inode. If False, the file is
moved, then file.replace writes to a new inode (mimicking prior
behavior). Default is True.

lorengordon committed:
Update file.replace to be hardlink-friendly, fixes saltstack#20969

This commit introduces the parameter `preserve_inode` to control the
behavior of file.replace. If `True`, the file is *copied*, then
file.replace writes to the original inode. If `False`, the file is
*moved*, then file.replace writes to a new inode (mimicking prior
behavior). Default is `True`.

thatch45 commented:
@lorengordon looks like this will not merge cleanly and has an error:
https://jenkins.saltstack.com/job/salt-pr-rs-cent7-n/1937/testReport/junit/unit.modules.file_test/FileReplaceTestCase/test_re_str_flags/

Can you please rebase and clean this up so we can get this merged in ? Thanks!

lorengordon commented:
@thatch45, yep, will do, no problem. I knew there would be a merge conflict with the symlink awareness patch.

lorengordon commented:
Don't merge this one just yet, please, still doing some testing. I'm having trouble understanding how jenkins is hitting that error, so trying another approach to managing the file handles...

lorengordon commented:
This PR should be good to go now. All the hard work was done in the 2015.2 branch, so with that merged forward, I just rebased my branch and added the `preserve_inode` parameter so that file.replace wouldn't clobber inodes and cause confusion with hard links.

thatch45 commented:
Thanks @lorengordon !
